students/jesse_miller/session09/cli_main.py

```diff
@@ -0,0 +1,171 @@
+#!/usr/bin/env python3
+'''
+This is the command interface for the mailroom.  Now that I know that donor
+manipulations work correctly (At least according to pytest.  I mean, I wrote
+the tests so the likelihood of human error here is high, you know?).
+'''
+#import os
+import sys
+#import datetime
+from donor_models import DonorCollection, Donor
+'''
+Module imports
+'''
+
+#pylint: disable=C0103
+alms = DonorCollection()
+'''
+Because my sense of humor is odd, that's why.  Also setting an object
+'''
+
+class MailRoom:
+    '''
+    All the calls and commands will be through this class.  Manipulation of
+    donor information is done in donor_models.py, and all of the execution for
+    this subprogram is from mailroom_oo.py
+    '''
+    @staticmethod
+    def list_donors():
+        """
+        Lists existing donors.
+        """
+        print(f"\n{'-'*15}\nList of Donors:\n{'-'*15}")
+        current = alms.list_donor()
+        for donor in current:
+            print(donor)
+        print(f"{'-'*15}\n")
+
+    @staticmethod
+    def print_report():
+        """
+        Prints a summary report of all donors.
+        """
+        headers = ["Donor Name", "Total Given", "Times Donated", "Average Gift"]
+        print(f"\n{'-'*80}\n{{:17}} | {{:<19}} | {{:<15}} | {{:<19}}\n{'-'*80}"\
+        .format(headers[0], headers[1], headers[2], headers[3]))
+
+        donor_data = alms.create_report()
+        for row in donor_data:
+            print('{:17} | ${:<18,.2f} | {:<15} | ${:<16,.2f}'.format\
+            (row[0], row[1], row[2], row[3]))
+        print(f"{'-'*80}\n")
+
+
+    @staticmethod
+    def donor_mail_choice():
+        '''
+        This section allows the user to mail a donor
+        '''
+        current_donor = ''
+        MailRoom.list_donors()
+        try:
+            donor = str(input('Who would you like to mail (all for all): '))
+            if alms.find_donor(donor):
+                MailRoom.mail_send_one(donor)
```

Comment:
Note that _donor_ here is the string value you're getting from the your _input_ call on line 62. You're then calling _mail_send_one()_, which in turn calls _Donor.letter_template(donor)_. 

Now, _letter_template(self)_ is supposed to be an instance method but the way you're calling it, it's behaving like a function that takes one parameter, _self_, which you're loading with _donor_, a string value (that's the reason for the "str object has no attribute name'). The way _letter_template_ is configured, it should be a Donor object, which would have a _name_ attribute.

Reply:
Another consideration to make is that _letter_template_ has no place in _donor_models.py_, because it deals with data visualization, not with data structure. _letter_template_ should instead be in _cli_main.py_.

One alternative would be to get _donor_dict_ from _cli_main.py_ and iterate through it, printing the thank you letter for each donor, sum(donations).

Another one would be to have a new instance method in _donor_models.py_ that does something like this:

def get_donor_total_donations(donor_name):
    return sum(self.donors_dict[current_donor])

Then you can call it from _cli_main.py_ when printing out thank you letters from each donor.